### PR TITLE
Paginate persisted cluster state

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we allocate a contiguous chunk of memory for the global metadata
each time we write it to disk. The size of this chunk is unbounded and
in practice it can be pretty large. This commit splits the metadata
document up into pages (1MB by default) that are streamed to disk at
write time, bounding the memory usage of cluster state persistence.
Since the memory usage is now bounded we can allocate a single buffer up
front and re-use it for every write.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 1]
```diff
@@ -580,6 +581,12 @@ private static Settings getRandomNodeSettings(long seed) {
             int retryTimeoutSeconds = RandomNumbers.randomIntBetween(random, 0, 60);
             builder.put(TransportReplicationAction.REPLICATION_RETRY_TIMEOUT.getKey(), timeValueSeconds(retryTimeoutSeconds));
         }
+        if (random.nextInt(10) == 0) {
+            builder.put(
+                PersistedClusterStateService.DOCUMENT_PAGE_SIZE.getKey(),
+                new ByteSizeValue(RandomNumbers.randomIntBetween(random, rarely() ? 10 : 100, 1000))
+            );
+        }
```
Comment (henningandersen) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:589]: I wonder if we should apply this more often but use between(10, 1000000) to allow for more variety of cases?

```suggestion
        if (randomBoolean()) {
            builder.put(
                PersistedClusterStateService.DOCUMENT_PAGE_SIZE.getKey(),
                new ByteSizeValue(RandomNumbers.randomIntBetween(random, rarely() ? 10 : 100, 1000000))
            );
        }
```

**Review by henningandersen — APPROVED** [review]
LGTM.

One question: I think we should never be able to see a downgraded node exercise this code since reading this happens after the `NodeEnvironment` reads node metadata. Just wanted to confirm that with you.

File: server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java [hunk 2]
```diff
@@ -511,13 +509,27 @@ private OnDiskState loadOnDiskState(Path dataPath, DirectoryReader reader) throw
...
+        IndexSearcher indexSearcher,
+        String type,
+        CheckedConsumer<BytesReference, IOException> bytesReferenceConsumer
+    ) throws IOException {
 
         final Query query = new TermQuery(new Term(TYPE_FIELD_NAME, type));
         final Weight weight = indexSearcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 0.0f);
         logger.trace("running query [{}]", query);
 
+        final Map<String, PaginatedDocumentReader> documentReaders = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java:531]: This runs fairly early within `Node#start` so there won't be much else happening in the node yet, and later on we will need to completely serialize the cluster state in memory anyway, so I think it should be fine. Also I think we'll be in trouble for other reasons sooner if index metadata exceeds 1MB when compressed and serialized. I added a comment in 16d6471e8cf.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java (hunk 1 above)
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: I think the more interesting corner cases are mostly when paginating harder; simply increasing the limit to 1000000 means that these interesting cases get much much less likely, but I introduced a variety of limits in ade43436fd8.

**Pull request merged**